File: ocaml/bob/bob.ml
open Base
open String

let response_for s =
  let s = s |> strip in
  let has_alpha s = s |> exists ~f:Char.is_alpha in
  if is_empty s then "Fine. Be that way!"
  else if has_alpha s && uppercase s = s then
    if s |> is_suffix ~suffix:"?" then "Calm down, I know what I'm doing!"
    else "Whoa, chill out!"
  else if s |> is_suffix ~suffix:"?" then "Sure."
  else "Whatever."

File: objective-c/bob/Bob.m
#import "Bob.h"

@implementation Bob
- (NSString *)hey: (NSString *)s {
    unsigned long last_idx = s.length - 1;
    nows = [[s lowercaseString]
              stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
    bool allcaps = ([[s uppercaseString] isEqualToString:s] && ![[s lowercaseString] isEqualToString:s]);
    if (nows.length == 0)
        return @"Fine. Be that way!";
    if (allcaps){
        return ([nows characterAtIndex:last_idx] == '?') ? @"Calm down, I know what I'm doing!" : @"Whoa, chill out!";
    }
    return ([nows characterAtIndex:last_idx] == '?' ? @"Sure" : @"Whatever";
}
@end
